File: server/src/db/rooms.py
from sqlite3 import Cursor

from src.db.util import read_sql_init_file
from src.models import Room
from src.logging import get_logger

logger = get_logger(__name__)


def init_rooms(cursor: Cursor):
    logger.info("Initializing rooms table")
    init_query = read_sql_init_file("rooms")
    try:
        cursor.execute(init_query)
    except Exception as e:
        logger.error('Failed to initialize rooms table', {'error': str(e)})
        raise e


def create_room(cursor: Cursor, room: Room):
    logger.info('Creating room', {'room': room})
    try:
        cursor.execute(
            """
            INSERT INTO rooms (name)
            VALUES (?)
            """,
            (room.name,),
        )
    except Exception as e:
        logger.error('Failed to create room', {'room': room, 'error': str(e)})
        raise e
    room.id = cursor.lastrowid
    logger.debug('Room created', {'room': room})


def update_room(cursor: Cursor, room: Room):
    logger.info('Updating room', {'room': room})
    try:
        cursor.execute(
            """
            UPDATE rooms
            SET name = ?
            WHERE id = ?
            """,
            (room.name, room.id),
        )
    except Exception as e:
        logger.error('Failed to update room', {'room': room, 'error': str(e)})
        raise e
    logger.debug('Room updated', {'room': room})


def delete_room(cursor: Cursor, room_id: int):
    logger.info('Deleting room', {'room_id': room_id})
    try:
        cursor.execute(
            """
            DELETE FROM rooms WHERE id = ?
            """,
            (room_id,),
        )
    except Exception as e:
        logger.error('Failed to delete room', {'room_id': room_id, 'error': str(e)})
        raise e
    logger.debug('Room deleted', {'room_id': room_id})


def find_room_by_id(cursor: Cursor, room_id: int) -> Room | None:
    logger.debug('Finding room by id', {'room_id': room_id})
    try:
        cursor.execute(
            """
            SELECT * FROM rooms WHERE id = ?
            """,
            (room_id,),
        )
        room = cursor.fetchone()
    except Exception as e:
        logger.error('Failed to find room by id', {'room_id': room_id, 'error': str(e)})
        raise e
    logger.debug('Room found', {'room': room})
    try:
        if room:
            return Room(**room)
        return None
    except Exception as e:
        logger.error('Failed to create room object', {'room': room, 'error': str(e)})
        raise e


def list_rooms(cursor: Cursor):
    logger.debug('Listing rooms')
    try:
        cursor.execute(
            """
            SELECT * FROM rooms
            """
        )
        rooms = cursor.fetchall()
    except Exception as e:
        logger.error('Failed to list rooms', {'error': str(e)})
        raise e
    logger.debug('Rooms listed', {'rooms': rooms})
    try:
        if rooms:
            return [Room(**room) for room in rooms]
        return []
    except Exception as e:
        logger.error('Failed to create room objects', {'rooms': rooms, 'error': str(e)})
        raise e

File: server-golang/db/rooms.go
// db/rooms.go
package db

import (
	"database/sql"
	"log"

	"github.com/topper3418/led-hub/golang-server/models" // Assume models import path.
)

func InitRooms(db *sql.DB) error {
	log.Println("Initializing rooms table")
	query, err := ReadSQLInitFile("rooms")
	if err != nil {
		return err
	}
	_, err = db.Exec(query)
	if err != nil {
		log.Printf("Failed to initialize rooms table: %v", err)
		return err
	}
	return nil
}

func CreateRoom(db *sql.DB, room *models.Room) error {
	log.Printf("Creating room: %+v", room)
	res, err := db.Exec("INSERT INTO rooms (name) VALUES (?)", room.Name)
	if err != nil {
		log.Printf("Failed to create room: %v", err)
		return err
	}
	id, err := res.LastInsertId()
	if err != nil {
		return err
	}
	room.ID = int(id)
	return nil
}

func UpdateRoom(db *sql.DB, room *models.Room) error {
	log.Printf("Updating room: %+v", room)
	_, err := db.Exec("UPDATE rooms SET name = ? WHERE id = ?", room.Name, room.ID)
	if err != nil {
		log.Printf("Failed to update room: %v", err)
		return err
	}
	return nil
}

func DeleteRoom(db *sql.DB, roomID int) error {
	log.Printf("Deleting room: %d", roomID)
	_, err := db.Exec("DELETE FROM rooms WHERE id = ?", roomID)
	if err != nil {
		log.Printf("Failed to delete room: %v", err)
		return err
	}
	return nil
}

func FindRoomByID(db *sql.DB, roomID int) (*models.Room, error) {
	log.Printf("Finding room by id: %d", roomID)
	room := &models.Room{}
	err := db.QueryRow("SELECT id, name FROM rooms WHERE id = ?", roomID).Scan(&room.ID, &room.Name)
	if err == sql.ErrNoRows {
		return nil, nil
	}
	if err != nil {
		log.Printf("Failed to find room by id: %v", err)
		return nil, err
	}
	return room, nil
}

func ListRooms(db *sql.DB) ([]*models.Room, error) {
	log.Println("Listing rooms")
	rows, err := db.Query("SELECT id, name FROM rooms")
	if err != nil {
		log.Printf("Failed to list rooms: %v", err)
		return nil, err
	}
	defer rows.Close()
	var rooms []*models.Room
	for rows.Next() {
		room := &models.Room{}
		if err := rows.Scan(&room.ID, &room.Name); err != nil {
			return nil, err
		}
		rooms = append(rooms, room)
	}
	return rooms, nil
}
